staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go

```diff
@@ -844,6 +845,177 @@ func TestNamespaceScopedWatch(ctx context.Context, t *testing.T, store storage.I
 	}
 }
 
+// RunOptionalTestWatchDispatchBookmarkEvents tests whether bookmark events are sent.
+// This feature is currently implemented in watch cache layer, so this is optional.
+//
+// TODO(#109831): ProgressNotify feature is effectively implementing the same
+//
+//	functionality, so we should refactor this functionality to share the same input.
+func RunTestWatchDispatchBookmarkEvents(ctx context.Context, t *testing.T, store storage.Interface) {
+	key, storedObj := testPropagateStore(ctx, t, store, &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: "test-ns"}})
+	startRV := storedObj.ResourceVersion
+
+	tests := []struct {
+		name                string
+		timeout             time.Duration
+		expected            bool
+		allowWatchBookmarks bool
+	}{
+		{ // test old client won't get Bookmark event
+			name:                "allowWatchBookmarks=false",
+			timeout:             3 * time.Second,
+			expected:            false,
+			allowWatchBookmarks: false,
+		},
+		{
+			name:                "allowWatchBookmarks=true",
+			timeout:             3 * time.Second,
+			expected:            true,
+			allowWatchBookmarks: true,
+		},
+	}
+
+	for i, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			pred := storage.Everything
+			pred.AllowWatchBookmarks = tt.allowWatchBookmarks
+			ctx, cancel := context.WithTimeout(context.Background(), tt.timeout)
+			defer cancel()
+
+			watcher, err := store.Watch(ctx, key, storage.ListOptions{ResourceVersion: startRV, Predicate: pred})
+			if err != nil {
+				t.Fatalf("Unexpected error: %v", err)
+			}
+			defer watcher.Stop()
+
+			// Create events of pods in a different namespace
```

Comment:
the watch is only for the specific key we created, right? so it's not necessary / a tiny bit misleading for this to be in a different namespace, would be sufficient to be another key?

Reply:
Well - yes and no.
Currently there are two separate mechanisms that are implemented:
- "progress notify" - it's an etcd feature that where etcd sends a bookmark with current (global) RV if no objects within the scope of the watch changed in the last N seconds
- "watch bookmarks" - it's a watchcache feature that sends a bookmark with its current RV (that in general is propagated only via watch of objects of a given type) [which may include the "progress notify events" if that feature is enabled

Given that those are currently two separate features, the test is currently only testing the second case (which doesn't need to enable progress-notify events).
The TODO in L851-852 is a desire to consolidate these two features a bit more and once we do that your comment would be a valid one - but for now, I think it's better to leave as is.